### PR TITLE
Config: Port EXI device and SI device settings to new config system.

### PR DESCRIPTION
**Pull request by AdmiralCurtiss**
Separate PR here because of this snippet of code that I'm not sure how to translate correctly into the new system:

```
    for (unsigned int i = 0; i < SerialInterface::MAX_SI_CHANNELS; ++i)
    {
      int source;
      controls_section->Get(fmt::format("PadType{}", i), &source, -1);
      if (source >= SerialInterface::SIDEVICE_NONE && source < SerialInterface::SIDEVICE_COUNT)
      {
        StartUp.m_SIDevice[i] = static_cast<SerialInterface::SIDevices>(source);
        config_cache.bSetPads[i] = true;
      }
    }
```

This happens while parsing GameINIs and basically means 'if `PadTypeN` is set to a valid value, override `SIDeviceN` with that, otherwise keep the value set by `SIDeviceN`'. I can and have added these to `GetINIToLocationMap()`, but that's not quite the same behavior -- if you set it to an invalid value, Dolphin will now use `SIDEVICE_NONE` rather than the current `SIDeviceN` value. Are we fine with this? If not, how do we adjust the code so that previous behavior still applies?

### CODE REVIEW

File: Source/Core/Core/HW/EXI/EXI.cpp [hunk 1]
```diff
@@ -101,7 +103,9 @@ void Init()
     AddMemoryCards(i);
 
   g_Channels[0]->AddDevice(EXIDEVICE_MASKROM, 1);
-  g_Channels[0]->AddDevice(SConfig::GetInstance().m_EXIDevice[2], 2);  // Serial Port 1
+  g_Channels[0]->AddDevice(
+      static_cast<ExpansionInterface::TEXIDevices>(Config::Get(Config::MAIN_SERIAL_PORT_1)),
+      2);  // Serial Port 1
```
Comment (JosJuice) [inline Source/Core/Core/HW/EXI/EXI.cpp:108]: This comment is redundant now (and makes the line wrapping kind of ugly).

**Comment by JosJuice** [comment]
I would be fine with this new PadType behavior.

File: Source/Core/Core/HW/SI/SI_DeviceGCController.cpp [hunk 2]
```diff
@@ -33,6 +34,14 @@ CSIDevice_GCController::CSIDevice_GCController(SIDevices device, int device_numb
   m_origin.origin_stick_y = GCPadStatus::MAIN_STICK_CENTER_Y;
   m_origin.substick_x = GCPadStatus::C_STICK_CENTER_X;
   m_origin.substick_y = GCPadStatus::C_STICK_CENTER_Y;
+
+  m_config_changed_callback_id = Config::AddConfigChangedCallback([this] { RefreshConfig(); });
+  RefreshConfig();
+}
+
+CSIDevice_GCController ::~CSIDevice_GCController()
```
Comment (JosJuice) [inline Source/Core/Core/HW/SI/SI_DeviceGCController.cpp:42]: Note sure why lint didn't catch this, but:

```suggestion
CSIDevice_GCController::~CSIDevice_GCController()
```

**Review by JosJuice — COMMENTED** [review]
There's a lot of casting going on in this PR. Is there any particular reason why the settings are defined as `int`s instead of the relevant enum type?

**Comment by AdmiralCurtiss** [comment]
I was somehow under the impression that you couldn't forward-declare an old C-style enum, but nope, seems to work fine. I've updated the settings to use the enum directly and removed the casts.

**Comment by leoetlino** [comment]
For the record, forward declaring an enum is fine as long as you specify the underlying type (which is only possible starting from C++11). Forward declaring an enum class without specifying the underlying type is okay too (int is the default).

**Review by JosJuice — APPROVED** [review]

**Pull request merged**